src/PactNet.Native/MessageScenarioBuilder.cs

```diff
@@ -0,0 +1,113 @@
+using System;
+using System.Collections.Generic;
+
+namespace PactNet.Native
+{
+    /// <summary>
+    /// Defines the message scenario builder
+    /// </summary>
+    public class MessageScenarioBuilder : IMessageScenarioBuilder
+    {
+        /// <summary>
+        /// Handles to create a new scenario with scenario builder
+        /// </summary>
+        public static IMessageScenarioBuilder NewScenario => new MessageScenarioBuilder();
+
+        /// <summary>
+        /// temporary description
+        /// </summary>
+        private string _descriptionAdded;
+
+        /// <summary>
+        /// tag to know if we are creating a scenario
+        /// </summary>
+        private bool _settingScenario;
+
+        private MessageScenarioBuilder()
+        {
+            Scenarios.All ??= new Dictionary<string, Func<dynamic>>();
+        }
+
+        /// <inheritdoc />
+        public IMessageScenarioBuilder WhenReceiving(string description)
+        {
+            try
+            {
+                ValidateDescription(description);
```

Comment:
Yeah, mentioned elsewhere in the review but the request/response API means that this is unnecessary because it's designed using the type-state pattern. Instead of having to validate that the user called the API in a valid order, you split the interfaces up into the various states and then add methods to each to transition to the next state.

In the request/response API the states are effectively:

- Uninitialised (an interaction hasn't been defined yet)
- Request (defining the request)
- Response (defining the response)
- Verifying (the interaction is built and can now be verified)

That's why there's separate interfaces for `IRequestBuilder` and `IResponseBuilder`. The only way to get to the response is to first create a request, and then use the `WillRespondWith` method to move to the response configuration state. This prevents the user from doing invalid things, and thus removes the need for this kind of validation.